## <center>Introdução à Linguagem Python</center>
###  <center>Projeto de ensino - IFB</center>
---
## <center>6- Mais Iterações</center>
##### <center>Prof.: Bruno V. Ribeiro

Há um outro comando `Python` que pode ser usado para construir uma iteração. Ele é o comando `while`. O `while` fornece um mecanismo muito mais geral para a iteração. Semelhante ao comando `if`, ele usa uma expressão booleana para controlar o fluxo de execução. O corpo do `while` será repetido enquanto a expressão booleana de controle for avaliada como `True` (**while** = enquanto em inglês).

Abaixo temos uma representação do fluxo do comando `while`.

<img src=https://panda.ime.usp.br/pensepy/static/pensepy/_images/while_flow.png>

## Leitura e implementação do `while`

O comando `while` pode ser "lido" como se fosse uma frase (lendo "enquanto" no lugar do `while`). Sua implementação é direta e segue as normas de bloco de execução do `if` e do `for`. Antes de escrevermos o código, vamos entender a chamada do comando:

- while (alguma condição testada):

    &nbsp;&nbsp;&nbsp; execute estas ações...
    
Simples assim. Vamos ver um exemplo de soma de números usando `while`:

In [0]:
# Somando números inteiros de 1 até o 'limite':
limite = 10

soma = 0
numero = 1

while numero <= limite:
    soma += numero
    numero += 1
    
print(soma)

55


Vamos entender cada linha acima:
  * Primeiro definimos o limite (até qual valor vamos somar);
  * Definimos `soma=0` pois antes de iniciar a contagem não há nada somado;
  * Definimos `numero = 1` para começar a soma do número 1.
  * Começamos o bloco do `while`:<space><space><space>
    * enquanto `numero<=limite`, faça <space><space><space><space><space>
    * some o valor de `numero` a `soma` <space><space><space><space><space>
    * incremente o valor de `numero` por 1 <space><space><space><space><space>
  * Ao acabarem as linhas dentro do bloco do `while` o código volta à primeira e testa a condição `numero<=limite`;
  * Se a condição ainda é satisfeita, continua dentro do bloco;
  * Se não, sai do bloco.

Formalmente, este é o fluxo de execução do comando `while`:

1. Avalie a condição, que deve resultar em `False` ou `True`.
2. Se o resultado for `False`, saia do comando `while` e continue o programa executando o próximo comando.
3. Se o resultado da condição for `True`, execute os comandos dentro do corpo do `while` e, ao final, retorne ao passo 1.

O corpo do `while` é constituído por todas as instruções abaixo do cabeçalho com a mesma tabulação.

Este tipo de fluxo é chamado de laço (`loop`) porque o terceiro passo volta para o topo, fechando o laço. Observe que se a condição é avaliada como `False` no início do laço, os comandos dentro do corpo nunca são executados.

O corpo do laço deve alterar o valor de uma ou mais variáveis para que a condição se torne `False` e faça o laço terminar. Caso contrário, o ciclo se repetirá para sempre. Isso é chamado de um laço infinito. Uma boa piada para os cientistas da computação são as instruções que acompanham alguns xampus: ensaboe, enxague, repita; que resulta em um laço infinito.

Vamos ver um exemplo simples de uso do `while` como ferramenta de controle. Voltemos ao nosso obeto em **MRU**. Vamos imaginar que todas as unidades estão do S.I. então não precisamos de nenhum tipo de correção por agora. Objeto possui velocidade `v=4` e posição inicial `x0=0`. Já sabemos, dos primeiros exercícios, que podemos criar uma sequência de código para calcular sua posição a qualquer instante. Podemos usar o `while` para descobrir em qual tempo ele atingirá uma certa posição `xf`, por exemplo (lógico que tudo isso pode ser feito facilmente com um lápis e um papel, mas vamos exercitar nossa programação!). Veja abaixo:

In [0]:
v = 4
x0 = 0
t = 0         # iniciando o tempo
x = x0        # iniciando a posição
# Esta é a informação mutável:
xf = 120

# Lembrem-se x = x0 + v*t
while x < xf:
    t += 1               # Passou um segundo!
    x = x0 + v*t         # Nosso objeto avançou no espaço

# Quando sairmos do while, nosso tempo (t) terá o valor correspondente a x = xf    
print("O objeto atinge a posição ", xf, " após ", t, " segundos.")

O objeto atinge a posição  120  após  30  segundos.


De fato: 120 = 0 + 4*30.

Note alguns detalhes: se mudarmos a posição inicial para `x0=1`, o que acontece? Vejamos

In [0]:
v = 4
# Mudei aqui!!!!
x0 = 1
t = 0         # iniciando o tempo
x = x0        # iniciando a posição
# Esta é a informação mutável:
xf = 120

# Lembrem-se x = x0 + v*t
while x < xf:
    t += 1               # Passou um segundo!
    x = x0 + v*t         # Nosso objeto avançou no espaço

# Quando sairmos do while, nosso tempo (t) terá o valor correspondente a x = xf    
print("O objeto atinge a posição ", xf, " após ", t, " segundos.")

O objeto atinge a posição  120  após  30  segundos.


A resposta não mudou. Mas, (120 = 1 + 4*30) me parece errado...

Este é um exemplo muuuuuuuuuuuuuuuuito simples (**MUITO**) de erro de precisão. Se resolvermos na mão o problema da última célula, chegamos à resposta t = 29,75 (acredito que too mundo consegue resolver, né?). O problema é que, aqui dentro do nosso `while`, o tempo é incrementado de 1 a cada iteração (veja a linha 11 da célula acima). Então, o que acontece é que 
* quando o t = 30 temos x = 1 + 4*30 = 121.
* Ao tentar entrar no while novamente, o programa testa: `x < 120`, ou seja 121 < 120?
* Como isso é falso, ele sai do laço com o valor de t = 30.

Como corrigimos isso? Usando passos de tempo menores. Podemos acrescentar uma variável `dt` ao nosso código para dizer ao programar o tamanho do paaso de tempo que queremos. Vamos fazer isso:

In [0]:
v = 4
x0 = 1
t = 0         # iniciando o tempo
x = x0        # iniciando a posição

# Esta é a informação mutável:
xf = 120

#Definindo passo de tempo
dt = 0.05

# Lembrem-se x = x0 + v*t
while x < xf:
    t += dt               # Passo de tempo igual a dt
    x = x0 + v*t         # Nosso objeto avançou no espaço

# Quando sairmos do while, nosso tempo (t) terá o valor correspondente a x = xf    
print("O objeto atinge a posição ", xf, " após ", "%.2f"%t, " segundos.")

O objeto atinge a posição  120  após  29.75  segundos.


Prontinho!

O comando `while` será muito útil para obtermos valores de quantidades dinâmicas (dependentes do tempo). Podemos, por exemplo, criar tabelas de valores a cada instante de tempo ao observarmos um objeto em movimento.

Vamos a outro exemplo. Imagine, agora, um objeto em MRUV (lembrem-se $S = S_0 + v_0 t + \frac{a}{2} t ^2$). Seja sua posição inicial `x0 = 0`, sua velocidade inicial `v0 = 20` e sua aceleração `a = -5`. Vamos fazer uma tabela acompanhando a evolução de sua posição ao passar o tempo por 10 segundos:

In [0]:
x0 = 0
v0 = 20
a = -5

# Definindo passo de tempo e tempo inicial
dt = 0.1
t = 0

# Inicializando a posição
x = x0
print("t = %.2f"%t , "-- xf = %.2f"%x)

# Tempo de observação
tf = 10


while t <= tf:
    t += dt
    x = x0 + v0*t + (a/2)*(t**2)
    print("t = %.2f"%t , "-- xf = %.2f"%x)
    


t = 0.00 -- xf = 0.00
t = 0.10 -- xf = 1.98
t = 0.20 -- xf = 3.90
t = 0.30 -- xf = 5.78
t = 0.40 -- xf = 7.60
t = 0.50 -- xf = 9.38
t = 0.60 -- xf = 11.10
t = 0.70 -- xf = 12.78
t = 0.80 -- xf = 14.40
t = 0.90 -- xf = 15.98
t = 1.00 -- xf = 17.50
t = 1.10 -- xf = 18.97
t = 1.20 -- xf = 20.40
t = 1.30 -- xf = 21.77
t = 1.40 -- xf = 23.10
t = 1.50 -- xf = 24.38
t = 1.60 -- xf = 25.60
t = 1.70 -- xf = 26.78
t = 1.80 -- xf = 27.90
t = 1.90 -- xf = 28.98
t = 2.00 -- xf = 30.00
t = 2.10 -- xf = 30.98
t = 2.20 -- xf = 31.90
t = 2.30 -- xf = 32.78
t = 2.40 -- xf = 33.60
t = 2.50 -- xf = 34.38
t = 2.60 -- xf = 35.10
t = 2.70 -- xf = 35.78
t = 2.80 -- xf = 36.40
t = 2.90 -- xf = 36.98
t = 3.00 -- xf = 37.50
t = 3.10 -- xf = 37.98
t = 3.20 -- xf = 38.40
t = 3.30 -- xf = 38.77
t = 3.40 -- xf = 39.10
t = 3.50 -- xf = 39.38
t = 3.60 -- xf = 39.60
t = 3.70 -- xf = 39.78
t = 3.80 -- xf = 39.90
t = 3.90 -- xf = 39.98
t = 4.00 -- xf = 40.00
t = 4.10 -- xf = 39.97
t = 4.20 -- xf = 39.90
t = 4.30 -- xf = 

Uma informação legal que podemos pedir ao código é para encontrar o **ponto de inflexão**, ou seja, quando a velocidade passou a ser negativa, ou seja de novo, quando a posição começou a diminuir.
Existem várias maneiras de fazer isso. Aqui vai uma delas:


In [0]:
x0 = 0
v0 = 20
a = -5

# Definindo passo de tempo e tempo inicial
dt = 0.1
t = 0

# Inicializando a posição
x = x0
print("t = %.2f"%t , "-- xf = %.2f"%x)

# Tempo de observação
tf = 10


while t <= tf:
    t += dt
    xa = x
    x = x0 + v0*t + (a/2)*(t**2)
    if x < xa:
        print("O ponto de inflexão é o x = %.2f"%xa)
        print("Acontece após t = %.2f"%(t-dt))
        break
    print("t = %.2f"%t , "-- xf = %.2f"%x)

t = 0.00 -- xf = 0.00
t = 0.10 -- xf = 1.98
t = 0.20 -- xf = 3.90
t = 0.30 -- xf = 5.78
t = 0.40 -- xf = 7.60
t = 0.50 -- xf = 9.38
t = 0.60 -- xf = 11.10
t = 0.70 -- xf = 12.78
t = 0.80 -- xf = 14.40
t = 0.90 -- xf = 15.98
t = 1.00 -- xf = 17.50
t = 1.10 -- xf = 18.97
t = 1.20 -- xf = 20.40
t = 1.30 -- xf = 21.77
t = 1.40 -- xf = 23.10
t = 1.50 -- xf = 24.38
t = 1.60 -- xf = 25.60
t = 1.70 -- xf = 26.78
t = 1.80 -- xf = 27.90
t = 1.90 -- xf = 28.98
t = 2.00 -- xf = 30.00
t = 2.10 -- xf = 30.98
t = 2.20 -- xf = 31.90
t = 2.30 -- xf = 32.78
t = 2.40 -- xf = 33.60
t = 2.50 -- xf = 34.38
t = 2.60 -- xf = 35.10
t = 2.70 -- xf = 35.78
t = 2.80 -- xf = 36.40
t = 2.90 -- xf = 36.98
t = 3.00 -- xf = 37.50
t = 3.10 -- xf = 37.98
t = 3.20 -- xf = 38.40
t = 3.30 -- xf = 38.77
t = 3.40 -- xf = 39.10
t = 3.50 -- xf = 39.38
t = 3.60 -- xf = 39.60
t = 3.70 -- xf = 39.78
t = 3.80 -- xf = 39.90
t = 3.90 -- xf = 39.98
t = 4.00 -- xf = 40.00
O ponto de inflexão é o x = 40.00
Acontece após t = 4.00


Tente acompanhar o código acima (na dúvida, me pergunte!). A ideia principal é que usamos uma variável auxiliar `xa` para armazenar a posição no passo anterior (antes de atualizar o valor de `x`) e comparar com `x` após atualizá-lo. Note, também, que usei o `break` para sair do laço depois que já achamos o ponto de inflexão.

## Exercício:

1) Repita o último exemplo (com os mesmos valores) mas, além da posição a cada passo de tempo, imprima também a velocidade (lembrem-se: $v = v_0 + a t$)

2) Tente encontrar o tempo em que o móvel volta a sua posição inicial. Lembre-se, a posição será descrita por uma parábola com concavidade negativa, então o móvel volta a sua posição original.

In [0]:
# Faça aqui o exercício 1.






In [0]:
# Faça aqui o exercício 2.






## Desafio:

Quando resolvemos o primeiro MRUV, obtivemos uma "tabela" de valores de posição para cada passo de tempo. Isso deixou a tela muito poluída. **SEM MUDAR** o valor de `dt`, faça o mesmo código que fizemos mas que imprima apenas as posições nos valores inteiros de `t`.

Ou seja, ao invés de vermos

t = 0.00 -- xf = 0.00

t = 0.10 -- xf = 1.98

t = 0.20 -- xf = 3.90

t = 0.30 -- xf = 5.78

t = 0.40 -- xf = 7.60

t = 0.50 -- xf = 9.38

t = 0.60 -- xf = 11.10

t = 0.70 -- xf = 12.78

t = 0.80 -- xf = 14.40

t = 0.90 -- xf = 15.98

t = 1.00 -- xf = 17.50

t = 1.10 -- xf = 18.97

t = 1.20 -- xf = 20.40

t = 1.30 -- xf = 21.77

t = 1.40 -- xf = 23.10

t = 1.50 -- xf = 24.38

t = 1.60 -- xf = 25.60

t = 1.70 -- xf = 26.78

t = 1.80 -- xf = 27.90

t = 1.90 -- xf = 28.98

t = 2.00 -- xf = 30.00


Quero ver apenas:

t = 0.00 -- xf = 0.00

t = 1.00 -- xf = 17.50

t = 2.00 -- xf = 30.00

In [0]:
# Faça aqui o desafio:

# Modifique o já escrevemos acima:
x0 = 0
v0 = 20
a = -5

# Definindo passo de tempo e tempo inicial
dt = 0.1
t = 0

# Inicializando a posição
x = x0
print("t = %.2f"%t , "-- xf = %.2f"%x)

# Tempo de observação
tf = 10


while t <= tf:
    t += dt
    x = x0 + v0*t + (a/2)*(t**2)
    print("t = %.2f"%t , "-- xf = %.2f"%x)

#### NOTA IMPORTANTE:

Notem que eu usei, em alguns lugares a notação (`"%.2f"%x`) para imprimir números. Isso diz ao Python para imprimir o valor de `x` com duas casas decimais. Veja:


In [0]:
x = 1/3
print(x)

0.3333333333333333


In [0]:
x = 1/3
print("%.2f"%x)

0.33


Vejam que isso muda o valor impresso, mas não altera o valor da variável!